plugins/inputs/vsphere/client.go

```diff
@@ -0,0 +1,75 @@
+package vsphere
+
+import (
+	"context"
+	"log"
+	"net/url"
+
+	"github.com/vmware/govmomi"
+	"github.com/vmware/govmomi/performance"
+	"github.com/vmware/govmomi/view"
+)
+
+// Client represents a connection to vSphere and is backed by a govmoni connection
+type Client struct {
+	Client *govmomi.Client
+	Views  *view.Manager
+	Root   *view.ContainerView
+	Perf   *performance.Manager
+	Valid  bool
+}
+
+// NewClient creates a new vSphere client based on the url and setting passed as parameters.
+func NewClient(u *url.URL, vs *VSphere) (*Client, error) {
+	tlsCfg, err := vs.TLSConfig()
+	if err != nil {
+		return nil, err
+	}
+	if vs.Username != "" {
+		log.Printf("D! Logging in using explicit credentials: %s", vs.Username)
+		u.User = url.UserPassword(vs.Username, vs.Password)
+	}
+	ctx := context.Background()
+
+	log.Printf("D! Creating client: %s", u.Host)
+	c, err := govmomi.NewClient(ctx, u, vs.InsecureSkipVerify)
+	if err != nil {
+		return nil, err
+	}
+
+	// Add certificate if we have it
+	//
+	if tlsCfg != nil && len(tlsCfg.Certificates) > 0 {
+		c.Client.SetCertificate(tlsCfg.Certificates[0])
+	}
+
+	c.Timeout = vs.Timeout.Duration
+	m := view.NewManager(c.Client)
+
+	v, err := m.CreateContainerView(ctx, c.ServiceContent.RootFolder, []string{}, true)
+	if err != nil {
+		return nil, err
+	}
+
+	p := performance.NewManager(c.Client)
+
+	return &Client{
```

Comment:
Done. No more ClientPool. There are only two connections: One used during metric collection and one used during discovery. To prevent a problem with stale connections, they are recreated every collection/discovery. Takes about 50ms, so we consider this a price worth paying to avoid having to write complex logic for cleaning up stale connections.